### PR TITLE
Fixes PageView.onPageChanged to only report changes in range

### PR DESCRIPTION
**Pull request by lukef**
This fix clamps the pixels used in the page calculation to be within min + max extents to avoid pages outside of the range of the children of a `PageView`.

### CODE REVIEW

**Comment by abarth** [comment]
Would you be willing to write a test for this change?  If you look in page_view_test.dart, you can see how it's done.  Basically, just add a NotificationListener and log the page values you see and then use `tester.scroll` to overscroll.  That will let us test that we don't get out-of-range notifications.

**Comment by abarth** [comment]
Or actually, you don't even need a NotificationListener.  You can just use onPageChanged.

**Comment by lukef** [comment]
Test added. Let me know if you want another test or want me to change it up. Sorry for the delay.

File: packages/flutter/test/widgets/page_view_test.dart [hunk 1]
```diff
@@ -15,20 +15,20 @@ void main() {
...
+                },
+                child: new Container(
+                  height: 200.0,
+                  decoration: const BoxDecoration(
+                    backgroundColor: const Color(0xFF0000FF),
+                  ),
+                  child: new Text(state),
+                ),
+              );
+            }).toList()
```
Comment (abarth) [inline packages/flutter/test/widgets/page_view_test.dart]: This will indent better if you add a trailing comma to this line.

Comment (lukef) [inline packages/flutter/test/widgets/page_view_test.dart]: Forgot one. Let me fix.

File: packages/flutter/test/widgets/page_view_test.dart [hunk 2]
```diff
@@ -265,19 +284,20 @@ void main() {
...
+              controller: controller,
+              itemCount: kStates.length,
+              itemBuilder: (BuildContext context, int index) {
+                return new Container(
+                  height: 200.0,
+                  color: index % 2 == 0 ? const Color(0xFF0000FF) : const Color(
+                          0xFF00FF00),
+                  child: new Text(kStates[index]),
+                );
+              }
```
Comment (abarth) [inline packages/flutter/test/widgets/page_view_test.dart]: dartfmt will indent this better if you put a trailing comma on this line.

File: packages/flutter/test/widgets/page_view_test.dart [hunk 3]
```diff
@@ -303,19 +323,24 @@ void main() {
   });
 
   testWidgets('PageView small viewportFraction', (WidgetTester tester) async {
+<<<<<<< HEAD
+    PageController controller = new PageController(viewportFraction: 1 / 8);
+=======
     final PageController controller = new PageController(viewportFraction: 1/8);
+>>>>>>> master
```
Comment (abarth) [inline packages/flutter/test/widgets/page_view_test.dart]: Looks like you've got some conflict markers to remove.

File: packages/flutter/test/widgets/page_view_test.dart [hunk 4]
```diff
@@ -303,19 +323,24 @@ void main() {
...
+              controller: controller,
+              itemCount: kStates.length,
+              itemBuilder: (BuildContext context, int index) {
+                return new Container(
+                  height: 200.0,
+                  color: index % 2 == 0 ? const Color(0xFF0000FF) : const Color(
+                          0xFF00FF00),
+                  child: new Text(kStates[index]),
+                );
+              }
```
Comment (abarth) [inline packages/flutter/test/widgets/page_view_test.dart]: trailing comma again.

File: packages/flutter/test/widgets/page_view_test.dart [hunk 5]
```diff
@@ -346,26 +372,49 @@ void main() {
...
```
Comment (abarth) [inline packages/flutter/test/widgets/page_view_test.dart]: ditto

File: packages/flutter/test/widgets/page_view_test.dart [hunk 6]
```diff
@@ -346,26 +372,49 @@ void main() {
...
     controller.jumpToPage(10);
     await tester.pump();
 
     expect(tester.getTopLeft(find.text('Hawaii')), const Point(-100.0, 0.0));
   });
-}
+
+  testWidgets('PageView does not report page changed on overscroll', (
+    WidgetTester tester) async {
+      PageController controller = new PageController(initialPage: kStates.length - 1);
```
Comment (abarth) [inline packages/flutter/test/widgets/page_view_test.dart]: final

File: packages/flutter/test/widgets/page_view_test.dart [hunk 7]
```diff
@@ -346,26 +372,49 @@ void main() {
...
+          controller: controller,
+          onPageChanged: (page) {
+            changeIndex = page;
+          }
+        );
+      }
+
+      await tester.pumpWidget(build());
+      controller.jumpToPage(kStates.length * 2);  // try to move beyond max range
+      expect(changeIndex, 0);   // change index should be zero, shouldn't fire onPageChanged
```
Comment (abarth) [inline packages/flutter/test/widgets/page_view_test.dart]: I'd put in an `await tester.pump()` and another `expect(changeIndex, 0);` just to make sure the notification doesn't wait for the pump.  (It doesn't, but that makes a more complete test.)

File: packages/flutter/test/widgets/page_view_test.dart [hunk 8]
```diff
@@ -346,26 +372,49 @@ void main() {
...
+      PageController controller = new PageController(initialPage: kStates.length - 1);
+      int changeIndex = 0;
+      Widget build() {
+        return new PageView(
+          children: kStates.map<Widget>((String state) => new Text(state))
+                  .toList(),
+          controller: controller,
+          onPageChanged: (page) {
+            changeIndex = page;
+          }
```
Comment (abarth) [inline packages/flutter/test/widgets/page_view_test.dart]: Missing trailing comma

**Comment by abarth** [comment]
LGTM

**Comment by abarth** [comment]
Thanks!

**Comment by lukef** [comment]
Fixed all commas, conflict marker (sorry, not sure how I missed that) and added additional code.

**Comment by abarth** [comment]
Looks good to me, but Travis has one more nit for you:

```
>>> RUNNING in .: bin/flutter analyze --flutter-repo
Analyzing 963 files...
[lint] Specify type annotations. (packages/flutter/test/widgets/page_view_test.dart, line 409, col 25)
```

File: packages/flutter/test/widgets/page_view_test.dart [hunk 9]
```diff
@@ -351,21 +375,48 @@ void main() {
...
+      (WidgetTester tester) async {
+    final PageController controller =
+        new PageController(initialPage: kStates.length - 1);
+    int changeIndex = 0;
+    Widget build() {
+      return new PageView(
+        children:
+            kStates.map<Widget>((String state) => new Text(state)).toList(),
+        controller: controller,
+        onPageChanged: (page) {
```
Comment (abarth) [inline packages/flutter/test/widgets/page_view_test.dart]: Travis wants to this to be `(int page)`

Comment (lukef) [inline packages/flutter/test/widgets/page_view_test.dart]: Fixed

**Comment by abarth** [comment]
A thing of beauty is a joy forever.

Thanks!

**Pull request merged**